### DISCUSSION TITLE
Performance Optimization Roadmap

### DISCUSSION THREAD

**Discussion by ayerofieiev-tt** [tenstorrent/pytorch2.0_ttnn#725] [General]
**High-Level Approach:**

- Compiler drives OP APIs to optimize performance  
  - via OP APIs memory usage, sharding, parallelization  
- OP improvements are continuous, not frozen.

### Phase 1a - Stop being a decelerator
* **Device OPs**: all OPs run on device, no fallbacks during the inference loop  
* **Pre-load weights**: once at the beginning, and not during the inference loop  
* **Proper perf measurement:** always hit kernel cache

### Phase 2a - Minimize OPs, Hardware Agnostic Optimizations
* Constant folding & propagation
* Minimize TransposeXY
* Simple OP fusion: common cases (Softmax, layernorm, conv+relu, mm+relu, etc)  

### Phase 2b: Optimize Batch size
* Increase tensor size to maximize core util, but not too large to cause too much spill to DRAM

### Phase 2c: Maximize Core Parallelization: TT-NN**
* Maximize core parallelization

### Phase 2d: Reduce Layout Transformation Overhead (OP-to-OP) TT-NN (TT-NN OP API dependency)
* Minimize tilize/untilize & pad/unpad: Toggle OP layout to minimize the intra-OP TMs

### Phase 3a: Maximize OP Sharding & Minimize Reshard OPs (TT-NN OP API dependency)
* Run OPs in sharded mode as much as possible  
* Minimize Re-shard OPs - keep sharding approach the same across OPs for as long as possible. 

### Phase 3b: Lazy-Tensor Squashing & OP reader Ingestion
* TT-NN tensor infrastructure, strides and stride accumulation  
* OP readers to understand the stride expressions 

### Phase 3c: Minimize spill to DRAM
* Simple heuristic**: query to keep intermediate tensors in L1 whenever possible  
* Graph schedule optimization**:   
  * Optimize the graph schedule to maximize tensor reuse from L1  
  * Optimize graph to maximize shard-chains

### Phase 3d: OP Perf Optimization (Torture) - OP microbenchmarking
* OP traces & perf optimization (Torture): get good perf on traces for OP shapes that most important. 

### Phase 3e: Optimize Large Resolution DRAM slicing & fusing:
* Chain Eltwise OPs onto DRAM-sliced OP parts
